app/controllers/posts_controller.rb

```diff
@@ -103,15 +98,20 @@ def post_params
   #or admin within the specific post, user will be redirected to the post they just tried to edit.
   #Otherwise, the user and/or the admin has the rights to edit/delete said post
   def authorize_user
+    action = params['action']
     post = Post.find(params[:id])
 
-    if current_user.moderator?
-      return current_user.moderator!
-    else
-      unless current_user == post.user || current_user.admin?
-        flash[:alert] = "You must be an admin to do that"
-        redirect_to [post.topic, post]
-      end
+    # A post owner should be able to do everything.
+    return true if post.user == current_user
+
+    # Only admins may destroy a post.
+    if action == 'destroy' && !current_user.admin?
+      flash[:alert] = "You must be an admin to do that"
+      redirect_to [post.topic, post]
+    # Only admins and moderators may update a post.
+    elsif action == 'update' && !(current_user.moderator? || current_user.admin?)
+      flash[:alert] = "You must be an admin or a moderator to do that"
+      redirect_to [post.topic, post]
     end
   end
```

Comment:
You had this chunk:

```ruby
if current_user.moderator?
  return current_user.moderator!
```

which essentially says: "if the currently signed-in user is a moderator, then make it a moderator, and return the result of this operation". It simply does not make sense.

The `authorize_user` filter is expected to either `return` (means it lets the request make it to the actual action), or redirect to some location, usually with a flash message for user feedback.